File: webapp/anti_patterns.py
"""CodeFluent — Structured output anti-pattern detection."""

import re

STRUCTURED_OUTPUT_PATTERNS: list[re.Pattern] = [
    re.compile(r"\boutput\s+(as|in)\s+json\b", re.IGNORECASE),
    re.compile(r"\brespond\s+(in|with)\s+json(\s+format)?\b", re.IGNORECASE),
    re.compile(r"\breturn\s+a?\s*json\s+(object|response|output)\b", re.IGNORECASE),
    re.compile(r"\bformat\s+(your\s+)?response\s+as\s+json\b", re.IGNORECASE),
    re.compile(r"\bgive\s+me\s+(the\s+)?(result|response|output|answer)\s+(as|in)\s+json\b", re.IGNORECASE),
    re.compile(r"\breply\s+(in|with)\s+json\b", re.IGNORECASE),
    re.compile(r"\bprovide\s+(the\s+)?(result|response|output|answer)\s+(as|in|using)\s+json\b", re.IGNORECASE),
    re.compile(r"\bjson\s+format\s+(only|please)\b", re.IGNORECASE),
    re.compile(r"\b(must|should|needs?\s+to)\s+(be|return|output)\s+(in\s+)?json\b", re.IGNORECASE),
]


def detect_structured_output_anti_pattern(prompts: list[str]) -> dict:
    """Scan prompts for structured output anti-patterns (requesting JSON format).

    Returns a dict with:
      - has_structured_output_antipattern: bool
      - structured_output_antipattern_count: int
      - structured_output_antipattern_indices: list[int]
    """
    indices: list[int] = []

    for i, prompt in enumerate(prompts):
        for pattern in STRUCTURED_OUTPUT_PATTERNS:
            if pattern.search(prompt):
                indices.append(i)
                break

    return {
        "has_structured_output_antipattern": len(indices) > 0,
        "structured_output_antipattern_count": len(indices),
        "structured_output_antipattern_indices": indices,
    }


def detect_anti_patterns(user_prompts: list[str]) -> dict:
    """Convenience wrapper for all anti-pattern detection.

    Currently delegates to structured output detection only.
    Provides an extension point for future anti-pattern types.
    """
    return detect_structured_output_anti_pattern(user_prompts)

File: vscode-extension/src/antiPatterns.ts
export type AntiPatternType = 'structured_output'

export interface AntiPatternResult {
  has_structured_output_antipattern: boolean
  structured_output_antipattern_count: number
  structured_output_antipattern_indices: number[]
}

export const STRUCTURED_OUTPUT_PATTERNS: RegExp[] = [
  /\boutput\s+(as|in)\s+json\b/i,
  /\brespond\s+(in|with)\s+json(\s+format)?\b/i,
  /\breturn\s+a?\s*json\s+(object|response|output)\b/i,
  /\bformat\s+(your\s+)?response\s+as\s+json\b/i,
  /\bgive\s+me\s+(the\s+)?(result|response|output|answer)\s+(as|in)\s+json\b/i,
  /\breply\s+(in|with)\s+json\b/i,
  /\bprovide\s+(the\s+)?(result|response|output|answer)\s+(as|in|using)\s+json\b/i,
  /\bjson\s+format\s+(only|please)\b/i,
  /\b(must|should|needs?\s+to)\s+(be|return|output)\s+(in\s+)?json\b/i,
]

export function detectStructuredOutputAntiPattern(prompts: string[]): AntiPatternResult {
  const indices: number[] = []

  for (let i = 0; i < prompts.length; i++) {
    const prompt = prompts[i]
    for (const pattern of STRUCTURED_OUTPUT_PATTERNS) {
      if (pattern.test(prompt)) {
        indices.push(i)
        break
      }
    }
  }

  return {
    has_structured_output_antipattern: indices.length > 0,
    structured_output_antipattern_count: indices.length,
    structured_output_antipattern_indices: indices,
  }
}

export function detectAntiPatterns(userPrompts: string[]): AntiPatternResult {
  return detectStructuredOutputAntiPattern(userPrompts)
}
